sycl/doc/design/ThinLTO.md

```diff
@@ -0,0 +1,149 @@
+# ThinLTO for SYCL
+
+This document describes the purpose and design of ThinLTO for SYCL.
+
+**NOTE**: This is not the final version. The document is still in progress.
+
+## Background
+
+With traditional SYCL device code linking, all user code is linked together
+along with device libraries into a single huge module and then split and
+processed by `sycl-post-link`. This requires sequential processing, has a large
+memory footprint, and differs from the linking flow for AMD and NVIDIA devices.
+
+## Summary
+
+SYCL ThinLTO will hook into the existing community mechanism to run LTO as part
+of device linking inside `clang-linker-wrapper`. We split the device images
+early at compilation time, and at link time we use ThinLTO's function importing
+feature to bring in the definitions for referenced functions. Only the new
+offload model is supported.
+
+## Device code compilation time changes
+
+Most of the changes for ThinLTO occur during device link time, however there is
+one major change during compilation (-c) time: we now run device code split
+during compilation instead of linking. The main reason for doing this is
+increased parallelization. Many compilation jobs can be run at the same time,
+but linking happens once total for the application. Device code split is
+currently a common source of performance issues.
+
+Splitting early means that the resulting IR after splitting is not complete, it
+still may contain calls to functions (user code and/or the SYCL device
+libraries) defined in other translation units.
+
+We rely on the assumption that all function definitions matching a declaration
+will be the same and we can let ThinLTO pull in any one.
+
+For example, let's start with user device code that defines a `SYCL_EXTERNAL`
+function `foo` in translation unit `tu_foo`. There is also another translation
+unit `tu_bar` that references `foo`. During the early device code splitting run
+of `tu_foo`, we may find that more than one of the resultant device images
+contain a definition for `foo`.
```

Comment:
Should we clarify the reason why `foo` could be duplicated? I'm not sure if that's obvious for a person who is not familiar with device code split details.

Also, we have a flag (originally introduced to support shared libraries) which allows to disable cloning of functions which are shared between device images produced by code splitting. We could probably use that mode for thinLTO to avoid unnecessary duplicates.

Reply:
will add this, thanks